server/src/main/java/org/elasticsearch/BuildVersion.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch;
+
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.internal.BuildExtension;
+import org.elasticsearch.plugins.ExtensionLoader;
+
+import java.util.ServiceLoader;
+
+public interface BuildVersion extends Writeable {
+    boolean onOrAfterMinimumCompatible();
+
+    boolean isFutureVersion();
+
+    // temporary
+    // TODO[wrb]: remove from PersistedClusterStateService
+    // TODO[wrb]: remove from security bootstrap checks
+    @Deprecated
+    default Version toVersion() {
+        return null;
+    }
+
+    static BuildVersion fromVersionId(int versionId) {
+        return ExtensionLoader.loadSingleton(ServiceLoader.load(BuildExtension.class))
+            .map(ext -> ext.fromVersionId(versionId))
+            .orElse(new DefaultBuildVersion(versionId));
+    }
+
+    static BuildVersion current() {
+        return ExtensionLoader.loadSingleton(ServiceLoader.load(BuildExtension.class))
+            .map(BuildExtension::currentBuildVersion)
+            .orElse(DefaultBuildVersion.CURRENT);
+    }
+
+    static BuildVersion empty() {
+        return ExtensionLoader.loadSingleton(ServiceLoader.load(BuildExtension.class))
+            .map(ext -> ext.fromVersionId(0))
+            .orElse(DefaultBuildVersion.EMPTY);
+    }
+
+    // only exists for NodeMetadata#toXContent
+    default int id() {
```

Comment:
Could BuildVersion have a toString that xcontent uses instead?

Reply:
I have been looking at this, but I am running into two things:

1. Round trip serialization -- when we get NodeMetadata by parsing XContent, we need an int value.
2. Backwards compatibility -- in GatewayMetastate, we use the NodeMetadata format to write NodeMetadata to disk. I think we have to write version as an `int` there, because otherwise old versions would not be able to understand the persisted node metadata as coming from a "future" version.

Please let me know if there's a way to get around these problems.

Reply:
I didn't realize the int version was so engrained in node metadata. Using a fixed value isn't the worst thing (should be hidden from users still).

> I think we have to write version as an int there, because otherwise old versions would not be able to understand the persisted node metadata as coming from a "future" version.

Is there any issue with using -1 as a "future" version?

Reply:
I think this should actually be abstract here, implemented as returning the version ID in DefaultBuildVersion, and implemented as returning a fixed value downstream.